tests/src/Simple/Threading/Threading.cs

```diff
@@ -758,13 +793,172 @@ public static void MutexMaximumReacquireCountTest()
+    }
+
+    [Fact]
+    public static void RepeatingWaitFiresUntilUnregistered()
+    {
+        var e0 = new AutoResetEvent(false);
+        var e1 = new AutoResetEvent(false);
+        var registered = ThreadPool.RegisterWaitForSingleObject(e0, (_, __) =>
+        {
+            e1.Set();
+        }, null, ThreadTestHelpers.UnexpectedTimeoutMilliseconds, false);
+        for (int i = 0; i < 4; ++i)
+        {
+            e0.Set();
+            Assert.True(e1.WaitOne(ThreadTestHelpers.ExpectedTimeoutMilliseconds));
+        }
+        var invalidWaitHandle = new InvalidWaitHandle();
+        registered.Unregister(invalidWaitHandle);
+        Thread.Sleep(50);
+        e0.Set();
+        Assert.False(e1.WaitOne(ThreadTestHelpers.ExpectedTimeoutMilliseconds));
+    }
+
+    [Fact]
+    public static void UnregisterEventSignaledWhenUnregistered()
+    {
+        var e0 = new AutoResetEvent(false);
+        var e1 = new AutoResetEvent(false);
+        var registered = ThreadPool.RegisterWaitForSingleObject(e0, (_, __) => {}, null, ThreadTestHelpers.UnexpectedTimeoutMilliseconds, true);
+        registered.Unregister(e1);
+        Assert.True(e1.WaitOne(ThreadTestHelpers.UnexpectedTimeoutMilliseconds));
+    }
+
+    [Fact]
+    public static void CanRegisterMoreThan64Waits()
+    {
+        RegisteredWaitHandle[] handles = new RegisteredWaitHandle[65];
+        for(int i = 0; i < 65; ++i) {
+            handles[i] = ThreadPool.RegisterWaitForSingleObject(new AutoResetEvent(false), (_, __) => {}, null, -1, true);
+        }
+        for(int i = 0; i < 65; ++i) {
+            handles[i].Unregister(null);
+        }
+    }
+
+    [Fact]
+    public static void StateIsPasssedThroughToCallback()
+    {
+        object state = new object();
+        AutoResetEvent e0 = new AutoResetEvent(false);
+        ThreadPool.RegisterWaitForSingleObject(new AutoResetEvent(true), (callbackState, _) =>
+        {
+            if(state == callbackState)
+            {
+                e0.Set();
+            }
+        }, state, 0, true);
+        Assert.True(e0.WaitOne(ThreadTestHelpers.UnexpectedTimeoutMilliseconds));
+    }
+
+    [Fact]
+    [OuterLoop]
+    public static void WaitWithLongerTimeoutThanWaitThreadCanStillTimeout()
+    {
+        AutoResetEvent e0 = new AutoResetEvent(false);
+        ThreadPool.RegisterWaitForSingleObject(new AutoResetEvent(false), (_, __) => e0.Set(), null, 21000, true);
+        Thread.Sleep(20000);
+        Assert.True(e0.WaitOne(ThreadTestHelpers.UnexpectedTimeoutMilliseconds));
+    }
+
+    [Fact]
+    public static void UnregisterCallbackIsNotCalledAfterCallbackFinishesIfAnotherCallbackOnSameWaitRunning()
+    {
+        AutoResetEvent e0 = new AutoResetEvent(false);
+        AutoResetEvent e1 = new AutoResetEvent(false);
+        RegisteredWaitHandle handle = ThreadPool.RegisterWaitForSingleObject(e0, (_, __) =>
+        {
+            Thread.Sleep(300);
```

Comment:
It looks like this test relies on this sleep duration not expiring before the test completes. Since any sleep can and often does take longer than the requested duration, I don't think there is enough gap to ensure the above. I would suggest increasing this duration and mark the test as [OuterLoop].